File: NHAY/NHAY_KMP.cpp
#include<stdio.h>
#include<iostream>
#include<vector>
#include<string>
using namespace std;
string needle;
string haystack;
string lesss    ;
vector<int> pi   ;
void Prefix_Function(string needle, long long int M){
    pi.push_back(0) ;
    long long int k = 0 ;
    for (long long int i = 1; i < M; i++){
        //cout << k << " " << i << " " << needle[k] << " " << needle[i] <<endl;
        while(needle[k] != needle[i]){
            if (!k)
                break   ;
            k = pi[k - 1]   ;
        }
        if (needle[k] == needle[i])
            k += 1  ;
        pi.push_back(k) ;
    }
    return  ;
}

void KMP(string needle, string haystack, long long int M){
    long long int N = haystack.length() ;
    long long int q = 0 ;
    int flag = 0    ;
    Prefix_Function(needle,M)   ;
    for (long long int i = 0; i < N; i++){
        while (needle[q] != haystack[i]){
            if (!q)
                break   ;
            q = pi[q - 1]   ;
        }
        if (needle[q] == haystack[i])
            q += 1  ;

        if (q == M){
            flag = 1    ;
            cout << i - M + 1 << "\n";
            q = pi[q - 1]   ;
        }
    }
    pi.clear()  ;
    if (flag)
        cout << "\n"   ;
    return ;
}
int main(){
    long long int n ;
    while(scanf("%lld", &n) == 1){
        getline(cin, lesss) ;
        getline(cin, needle);
        getline(cin, haystack)  ;
//        printf("%s %s", needle, haystack)   ;
//        cout << needle << " " << haystack   << endl;
        KMP(needle, haystack , n)   ;

    }
    return 0    ;
}

File: NHAY/NHAY_KMP.py
__author__ = "Ashwani Gautam"

import sys

def main():
    def Prefix_Function(needle, M):
        pi = [0] * M
        k = 0
        for i in xrange(1, M):
            while needle[k] <> needle[i]:
                if k == 0:
                    break
                k = pi[k - 1]
                
            if needle[k] == needle[i]:
                k += 1
                
            pi[i] = k
        print pi
        return pi

    def KMP(needle, haystack, M):
        N = len(haystack)
        pi = Prefix_Function(needle, M)
        q = 0
        flag = 0
        for i in xrange(N):
            while needle[q] <> haystack[i]:
                if q == 0:
                    break
                q = pi[q - 1]
                
            if needle[q] == haystack[i]:
                q += 1

            if q == M:
                flag = 1
                print i - M + 1
                q = pi[q - 1]
        if flag:
            print "\n"
    

    flag = 0
    for line in sys.stdin:
        if flag == 0:
            M = int(line.strip())
            flag = 1
            continue
        elif flag == 1:
            needle = line.strip()
            flag = 2
            continue
        else:
            haystack = line.strip()
            flag = 0
        KMP(needle, haystack, M)
if __name__ == "__main__":
    main()
